lib/Sema/TypeCheckEffects.cpp

```diff
@@ -3593,16 +3593,38 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
       if (const AnyTryExpr *tryExpr = dyn_cast<AnyTryExpr>(autoClosure->getSingleExpressionBody()))
         awaitInsertLoc = tryExpr->getSubExpr()->getStartLoc();
     }
+    
+    auto findShorthandOptionalBinding = [&](Expr *anchor) -> VarDecl * {
+      VarDecl *varDecl = nullptr;
+      if (!anchor->getType()->isOptional())
+        return nullptr;
+      anchor->forEachChildExpr([&](Expr *expr) -> Expr * {
+        if (isa<DeclRefExpr>(expr) && expr->isImplicit()) {
+          varDecl = dyn_cast<VarDecl>(expr->getReferencedDecl().getDecl());
+          return nullptr;
+        }
+        return expr;
+      });
+      return varDecl;
+    };
```

Comment:
I would tighten up this logic to look through all `ImplicitConversionExpr` layers instead of walking all children. I think even just the existence of an implicit `DeclRefExpr` is acceptable to conclude a shorthand optional binding. If not, the additional optional type condition would only get in the way of us knowing about it because there definitely is no case where an implicit `DeclRefExpr` cannot have optional type. Alternatively, a robust solution would be to implement `ASTWalker::walkToStmtConditionElementPre/Post`, and use that to store the condition element in the context object before descending into the implicit initializer expression ([`Context::InterpolatedString`](https://github.com/apple/swift/blob/5a546ed6b0c60ef522ec1898b63b8ec489a24542/lib/Sema/TypeCheckEffects.cpp#L2112) serves a similar purpose). `StmtConditionElement` has all the info needed to outline this case.

Reply:
I agree, that was somewhat the point on the previous comment, be more specific and robust on the check :)